explorer/exporters.py

```diff
@@ -0,0 +1,99 @@
+import json
+
+from django.http import HttpResponse
+from django.utils.module_loading import import_string
+
+from . import app_settings
+from .utils import csv_report, get_filename_for_title
+
+
+def get_exporter_class(format):
+    class_str = getattr(app_settings, 'EXPLORER_DATA_EXPORTERS')[format]
+    return import_string(class_str)
+
+
+class BaseExporter(object):
+
+    name = ''
+    content_type = ''
+
+    def __init__(self, query):
+        self.query = query
+
+    def to_response(self):
+        raise NotImplementedError
```

Comment:
What do you think of putting more logic into the base class? I think it would increase cohesion, and allow us to remove some orphaned functions from utils. Something like (totally untested):

```py
class BaseExporter(object):

    name = ''
    content_type = ''

    def __init__(self, query):
        self.query = query

    def get_data(self, **kwargs):
        raise NotImplementedError

    def get_filename(self):  # inlined utils.get_filename_for_title()
        valid_chars = '-_.() %s%s' % (string.ascii_letters, string.digits)
        filename = ''.join(c for c in self.query.title if c in valid_chars)
        filename = filename.replace(' ', '_')
        return filename

    def to_response(self, download=True, **kwargs):
        data = get_data(self.query, **kwargs)
        response = HttpResponse(data, content_type=self.content_type)
        response['Content-Length'] = len(data)
        if download:
            response['Content-Disposition'] = 'attachment; filename="%s.csv"' % self.get_filename()
        return response
```

Then CSVExporter just becomes (after inlining write_csv and csv_report from utils):

```py
class CSVExporter(BaseExporter):
    name = 'CSV'
    content_type = 'text/csv'

    def get_data(self, delim=settings.CSV_DELIMITER):
        try:
            res = self.query.execute_query_only()
        except DatabaseError as e:
            resp = cStringIO()
            return resp.write(str(e))  # consistent return type
        
        delim = '\t' if delim == 'tab' else delim
        csv_data = cStringIO()
        if PY3:
            writer = csv.writer(csv_data, delimiter=delim)
        else:
            writer = csv.writer(csv_data, delimiter=delim, encoding='utf-8')
        writer.writerow(res.headers)
        for row in res.data:
            writer.writerow([s for s in row])
        return csv_data.get_value()
```

This contains all of the gnarliness around specific formats to one method in each exporter, decoupling the HttpResonse bits from the format handling. You could imagine adding a `to_file()` method on the BaseExporter, for instance.

With all that said, someone smart once said "[don't make classes that end in -er](http://objology.blogspot.com/2011/09/one-of-best-bits-of-programming-advice.html)", so another option to consider is keeping the "response-building" methods on the views as you suggested in your comment, and instead adding `.to_<format>()` methods to the Query class itself. It doesn't have the same sex-appeal as the multiple-classes + exporter 'registry', but is in many ways a more SOLID approach.